### PR TITLE
Propagate assertions for more checked bounds

### PR DESCRIPTION
**Pull request by JosephTremoulet**
Generalize (and rename) the assertion/limit code for tracking array
lengths (which have until now been recognized by the VNFunc being
GT_ARR_LENGTH) to track all values that appear in the function as
length arguments to bounds chcks nodes.  Add a hashtable to identify that
set of value numbers, `m_checkedBoundVNs` to the VNStore object, and
populate it during value-numbering.  Wherever names were adjusted, the
generalization of "array length" is called "checked bound".

This allows the array bound check removal machinery to operate on span
bound checks as well.

### CODE REVIEW

**Comment by JosephTremoulet** [comment]
@briansull, @AndyAyersMS PTAL
/cc @dotnet/jit-contrib

This addresses #10050 with the exception that the value-number update that happens for compare nodes when one of their arguments is an array length that gets CSE'd does not apply in as many cases to spans, because we don't have the CSE candidate occurrence lists collected yet when we build the `optCseCheckedBoundMap`; that is left as a future possible improvement.

The `Performance/CodeQuality/Span/Indexer` test consists of several indexing patterns which we can use to compare the power of span bound check removal against array bound check removal.  The effect of this change (on top of #11207) is:

|      Sub-Test    |      Effect of this change     |     Effect of this change plus #10453    |    Effect of this change plus fix for `optCseCheckedBoundMap` issue  |
|-----------------|---------------------------------|-----------------------------------------|----------------------------------------------------------------------------------|
|     Indexer1    |     bound check removed; array parity   |     |     |
|     Indexer2   |      no change (CSE doesn't update VN)   |  bound check removed; array parity   |  bound check removed; array parity    |
|     Indexer3    |     bound checked removed; array parity   |     |     |
|     Indexer4    |     bound checked removed; array parity   |     |     |
|     Indexer5    |     no change (afraid write could clobber span length)   |  bound checked removed; array parity   |     |
|     Indexer6    |     no change (afraid write could clobber span length)   |  bound checked removed; array parity   |     |
|     ReadOnlyIndexer1    |     bound check removed; array parity   |     |     |
|     ReadOnlyIndexer2   |      no change (CSE doesn't update VN)   |  bound check removed; array parity   |  bound check removed; array parity    |
|     WriteViaIndexer1    |     no change (afraid write could clobber span length)   |  bound checks on data[idx] removed, not on data[0]   |     |
|     WriteViaIndexer2    |     no change, but already have array parity   |     |     |
|     KnownSizeArray    |     no change, but already have array parity   |     |     |
|     KnownSizeCtor    |     no change, but already have array parity   |     |     |
|     KnownSizeCtor2    |     no change, but already have array parity   |     |     |
|     SameIndex1    |     no change, but already have array parity   |     |     |
|     SameIndex2    |     no change (afraid write could clobber span length)   |  bound checked removed; array parity   |     |
|     CoveredIndex1    |     no change, but already have array parity   |     |     |
|     CoveredIndex2    |     no change, but already have array parity   |     |     |
|     CoveredIndex3    |     no change, but already have array parity   |     |     |      

No other diffs from this change (spans aren't used much in our test corpus)

File: src/jit/assertionprop.cpp [hunk 1]
```diff
@@ -1788,16 +1788,16 @@ AssertionInfo Compiler::optCreateJTrueBoundsAssertion(GenTreePtr tree)
...
         return index;
     }
-    // Cases where op1 holds the condition array length and op2 is 0.
-    // Loop condition like: "i < a.len == 0"
-    // Assertion: "i < a.len == false"
-    else if (vnStore->IsVNArrLenBound(vn) &&
+    // Cases where op1 holds the condition bound and op2 is 0.
+    // Loop condition like: "i < len == 0"
+    // Assertion: "i < len == false"
+    else if (vnStore->IsVNCompareCheckedBound(vn) &&
```
Comment (briansull) [inline src/jit/assertionprop.cpp]: The comment block reads better like this:


// Cases where op1 holds the upper bound and op2 is 0.
// Loop condition like: "i < bnd == 0"
// Assertion: "i < bnd == false"

Comment (AndyAyersMS) [inline src/jit/assertionprop.cpp]: If you change this also change the previous comment block to match.

Comment (AndyAyersMS) [inline src/jit/assertionprop.cpp]: And a few more below...

File: src/jit/valuenum.cpp [hunk 2]
```diff
@@ -7028,6 +7062,14 @@ void Compiler::fgValueNumberTree(GenTreePtr tree, bool evalAsgLhsInd)
                 excSet = vnStore->VNPExcSetUnion(excSet, vnStore->VNPExcVal(tree->AsBoundsChk()->gtArrLen->gtVNPair));
 
                 tree->gtVNPair = vnStore->VNPWithExc(vnStore->VNPForVoid(), excSet);
+
+                // Record non-constant value numbers that are used a the length argument to bounds checks, so that
```
Comment (mikedn) [inline src/jit/valuenum.cpp]: are used a**s** the length

**Comment by JosephTremoulet** [comment]
Updated per feedback (and to fix build issue)

**Comment by JosephTremoulet** [comment]
Perf lab run came back, the indexer tests with diffs do indeed show improvement:

|     Test         |      Improvement      |
|--------------|-----------------------|
| Indexer1      |             19%             |
|  Indexer3    |              18%             |
|  Indexer4     |              7%             |
|  ReadOnlyIndexer1  |      19%        |

**Comment by JosephTremoulet** [comment]
@dotnet-bot test this please

**Comment by JosephTremoulet** [comment]
test OSX10.12 x64 Checked Build and Test

**Pull request merged**